Fix race condition causing containers to fail to start. Fixes #6.

SharkWipf opened the pull request:
This change fixes host firewall race conditions causing containers to sometimes fail to start.
In order to avoid concurrent modification, iptables tracks locks in the file `/run/xtables.lock`.
In order for iptables inside a VM to respect the locking, it will need access to this file as well (https://github.com/moby/moby/issues/12547#issuecomment-306143160).

The downside of this implementation, as explained in above linked comment, is that, in some configurations, the lockfile may not exist when the container is started, and Docker will create it as a directory.
Sadly there's no reliable/non-hacky way to avoid this within the scope of this project that I know of, as any method to create this file would require root access, and the container user on the host is not likely to be root.

SharkWipf committed:
Fix race condition causing containers to fail to start. Fixes #6.

mitar commented:
Thanks. This looks awesome. But why not mount whole `/run` directory inside?

mitar commented on README.md:
```diff
@@ -10,3 +10,6 @@ containers external IP, iptables will be configured to route container's traffic
 And one more empty chain is created after this one for any additional custom rules you might want
 to add, named `AFTER_EXTERNAL_IP`.
+
+Please make sure `/run/xtables.lock` exists on the host before starting the container.
```
Can you explain why this mount is necessary?

SharkWipf replied:
Added some extra explanation in the latest commit

mitar commented on README.md:
````diff
@@ -1,7 +1,7 @@
 
 ```
-$ docker run --detach --net=host --cap-add=NET_ADMIN --cap-add=NET_RAW --volume /var/run/docker.sock:/var/run/docker.sock tozd/external-ip
+$ docker run --detach --net=host --cap-add=NET_ADMIN --cap-add=NET_RAW --volume /var/run/docker.sock:/var/run/docker.sock --volume /run/xtables.lock:/run/xtables.lock tozd/external-ip
````
This line is getting long, can you break it into multiple lines (using `\`)?

SharkWipf replied:
Done in the latest commit

SharkWipf committed:
Merge branch 'master' into master

SharkWipf commented:
Sorry for the delay, only just got back to this.

> Thanks. This looks awesome. But why not mount whole `/run` directory inside?

This would expose a lot more than necessary, potentially even sensitive data.  
This container already needs more access than most containers would need to do its job, and it could already potentially gain a lot of access to the host, but it's still good practice to not expose anything more than strictly necessary IMO.
On top of that, while this container doesn't currently run anything that uses /run directly that isn't already passed to the container, adding a broad mount like `/run` would increase the risk of conflicts in this directory, and potentially decrease portability.

I've updated and implemented the feedback in the latest commit in the PR, if there's anything else you'd like to see changed please let me know.

mitar commented:
> This would expose a lot more than necessary, potentially even sensitive data.

I mean, with access to `docker.sock` the container has literally root access on the host. So I am not sure if being precise about `/run` mount is really important and maybe it is just a security theater.

But having a file-only mount means that you have the whole issue with the file maybe not existing on the host.

mitar commented on README.md:
```diff
@@ -15,6 +16,12 @@ A chain named `EXTERNAL_IP` is created in the `nat` table into which all the rul
+preventing race conditions that can cause containers to fail to start.
+If this file does not exist, Docker will incorrectly create it as a directory, which may cause issues both on the host and with the container.
+
+=======
```
This is leftover from some bad merge?
